File: src/application/queries/get_detail_action_query_handler.py
from typing import Optional

from src.application.queries.query_handler import Query, QueryHandler
from src.application.queries.query_models.action_query_model import ActionQueryModel
from src.application.repositories.actions_repository import ActionsRepository


class GetDetailActionQuery(Query):
    def __init__(self, id_action: str):
        super(GetDetailActionQuery, self).__init__(name='GetDetailActionQuery')
        self.id_action = id_action


class GetDetailActionQueryHandler(QueryHandler):
    def __init__(self, actions_repository: ActionsRepository):
        self.actions_repository = actions_repository

    def handle(self, query: GetDetailActionQuery) -> Optional[ActionQueryModel]:
        return self.actions_repository.get_query_model(query.id_action)

File: src/application/queries/get-detail-action.query.handler.ts
import { Inject, Injectable } from '@nestjs/common'
import { Authentification } from 'src/domain/authentification'
import { Query } from '../../building-blocks/types/query'
import { QueryHandler } from '../../building-blocks/types/query-handler'
import { Action, ActionsRepositoryToken } from '../../domain/action'
import { ActionAuthorizer } from '../authorizers/authorize-action'
import { ActionQueryModel } from './query-models/actions.query-model'

export interface GetDetailActionQuery extends Query {
  idAction: string
}

@Injectable()
export class GetDetailActionQueryHandler extends QueryHandler<
  GetDetailActionQuery,
  ActionQueryModel | undefined
> {
  constructor(
    @Inject(ActionsRepositoryToken)
    private actionRepository: Action.Repository,
    private actionAuthorizer: ActionAuthorizer
  ) {
    super()
  }

  async handle(
    query: GetDetailActionQuery
  ): Promise<ActionQueryModel | undefined> {
    return this.actionRepository.getQueryModelById(query.idAction)
  }

  async authorize(
    query: GetDetailActionQuery,
    utilisateur: Authentification.Utilisateur
  ): Promise<void> {
    await this.actionAuthorizer.authorize(query.idAction, utilisateur)
  }

  async monitor(): Promise<void> {
    return
  }
}
